src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,298 @@
+    private Map<String, Integer>                RECURSIVE_TYPES = Maps.newConcurrentMap();
+
+    private final Map<String, Supplier<TypeTemplate>> MODDED_ENTITY_TYPES = Maps.newConcurrentMap();
+    private final Map<String, Supplier<TypeTemplate>> MODDED_BLOCK_ENTITY_TYPES = Maps.newConcurrentMap();
+
+    //Store some data we can not access later anymore.
+    private final String name;
+    private final Schema wrapped;
+
+    ForgeSchema(final int versionKey, final Schema parent, final Schema wrapped)
+    {
+        super(versionKey, parent);
+        this.wrapped = wrapped;
+
+        final int subVersion = DataFixUtils.getSubVersion(versionKey);
+        name = "V" + DataFixUtils.getVersion(versionKey) + (subVersion == 0 ? "" : "." + subVersion);
+
+        //Now with the vanilla data in the schema, lets build our first version again.
+        //Yes this executes the type resolve multiple times per schema (but it's not a big deal)
+        resetSchema();
+        rebuildSchema(
+          Collections.emptyMap(),
+          Collections.emptyMap()
+        );
+    }
+
+    /**
+     * Rebuilds the types, with support for the case that no types exists.
+     * (Which should never happen, but you never know!)
+     *
+     * @return The type map.
+     */
+    protected Map<String, Type<?>> buildTypes()
+    {
+        //Result map.
+        final Map<String, Type<?>> types = Maps.newHashMap();
+
+        //All the core templates into which recursion of data is possible.
+        final List<TypeTemplate> templates = Lists.newArrayList();
+
+        //Check all recursion types
+        for (final Map.Entry<String, Integer> entry : RECURSIVE_TYPES.entrySet())
+        {
+            //Generate a template for the type.
+            templates.add(DSL.check(entry.getKey(), entry.getValue(), getTemplate(entry.getKey())));
+        }
+
+        //Merge them all together into one root recursive type.
+        final Optional<TypeTemplate> availableTypeTemplate = templates.stream().reduce(DSL::or);
+        if (availableTypeTemplate.isEmpty())
+        {
+            //Check for its existence or nuke.
+            return Maps.newHashMap();
+        }
+        //Build a type family for our recursive types.
+        final TypeTemplate choice = availableTypeTemplate.get();
+        final TypeFamily family = new RecursiveTypeFamily(name, choice);
+
+        //Now process all other known types, taking care of recursion.
+        for (final String name : TYPE_TEMPLATES.keySet()) {
+            final Type<?> type;
+            final int recurseId = RECURSIVE_TYPES.getOrDefault(name, -1);
+            if (recurseId != -1)
+            {
+                type = family.apply(recurseId);
+            } else {
+                type = getTemplate(name).apply(family).apply(-1);
+            }
+            types.put(name, type);
+        }
+        return types;
+    }
+
+    public Set<String> types()
```

Comment:
Missing override annotation (on this method and a bunch of others in this class)? It also is probably worth adding a one line comment/javadoc for the methods that are carbon copies of super, in case things in DataFixerUpper change so as to make it easier to track.